### PR TITLE
OrganizationAccessActions

### PR DESCRIPTION
**Pull request by Colin-Lane**
- 2 simple solutions, lets see which one is liked more.
  @leogrim @ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala [hunk 1]
```diff
@@ -18,22 +18,43 @@ trait OrganizationAccessActions {
...
+          Future.successful(OrganizationFail.INVALID_PUBLIC_ID.asErrorResponse)
+      }
+    }
+  }
+
+  def OrganizationAccessCheck(id: PublicId[Organization], permissions: OrganizationPermission*) = new ActionFilter[MaybeUserRequest] {
+    def filter[A](input: MaybeUserRequest[A]): Future[Option[Result]] = Future.successful(lookupViewable(id, input, permissions))
+  }
+
+  private def lookupViewable[A](orgPubId: PublicId[Organization], input: MaybeUserRequest[A], permissions: Seq[OrganizationPermission]) = {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: Can we call these `requiredPermissions`?

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala [hunk 2]
```diff
@@ -18,22 +18,43 @@ trait OrganizationAccessActions {
...
     parseRequest(orgPubId, input) match {
       case Some((orgId, userIdOpt, authToken)) =>
         // Right now all organizations are publicly visible
-        val access = true // TODO: this ought to be removed when Leo gets to OrganizationAccessAction.byPermission
-        if (access) {
+        val memberPermissions = orgMembershipCommander.getPermissions(orgId, userIdOpt)
+        if (permissions forall (memberPermissions.contains)) {
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: I think that Scala has a `subsetOf` function, so `permissions subsetOf memberPermissions`. I had a few issues with type-checking the last time I tried to used it, but maybe it'll work here.

Comment (Colin-Lane) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: permissions is a seq. I could convert to a set and then use subsetOf. I considered doing that, but I am not sure I like it more.

**Comment by ryanpbrewster** [comment]
I prefer the `OrganizationAccessCheck` overall.

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: +1

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala [hunk 3]
```diff
@@ -18,22 +18,43 @@ trait OrganizationAccessActions {
...
+  case class OrganizationRequest[T](request: MaybeUserRequest[T], orgId: Id[Organization], authToken: Option[String], permissions: Set[OrganizationPermission]) extends WrappedRequest[T](request) {
+    def userIdOpt = request.userIdOpt
   }
 
-  private def lookupViewable[A](orgPubId: PublicId[Organization], input: MaybeUserRequest[A]) = {
+  def OrganizationAction(id: PublicId[Organization], permissions: OrganizationPermission*) = MaybeUserAction andThen new ActionFunction[MaybeUserRequest, OrganizationRequest] {
+    override def invokeBlock[A](request: MaybeUserRequest[A], block: (OrganizationRequest[A]) => Future[Result]): Future[Result] = {
+      parseRequest(id, request) match {
+        case Some((orgId, userIdOpt, authToken)) =>
+          // Right now all organizations are publicly visible
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: Let's just always add `VIEW_ORGANIZATION` to the set of explicitly required permissions

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: Better to make it a `Set` anyway, that's the real semantics (unique + unordered)

File: kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala [hunk 4]
```diff
@@ -18,22 +18,43 @@ trait OrganizationAccessActions {
...
+        val memberPermissions = orgMembershipCommander.getPermissions(orgId, userIdOpt)
+        if (permissions forall (memberPermissions.contains)) {
           None
         } else {
-          Some(Forbidden(Json.obj("error" -> "permission_denied")))
+          Some(OrganizationFail.INSUFFICIENT_PERMISSIONS.asErrorResponse)
         }
       case _ =>
-        Some(BadRequest(Json.obj("error" -> "invalid_id")))
+        Some(OrganizationFail.INVALID_PUBLIC_ID.asErrorResponse)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/shoebox/controllers/OrganizationAccessActions.scala]: Can we move this logic into the main method? There's only one, we don't need to factor it out.

**Pull request merged**